compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNotImplementedOverrideChecker.kt

```diff
@@ -0,0 +1,278 @@
+    private fun getBaseDeclarationsForFakeOverrides(
+        firClass: FirClass<*>,
+        contributedDeclarations: Collection<FirDeclaration>,
+        context: CheckerContext
+    ): Collection<FirCallableDeclaration<*>> {
+        val result = mutableListOf<FirCallableDeclaration<*>>()
+        val classScope = firClass.unsubstitutedScope(context)
+
+        fun checkFunctionSymbolAndAddToResult(originalSymbol: FirCallableSymbol<*>) {
+            if (shouldCreateFakeOverridden<FirSimpleFunction, FirNamedFunctionSymbol>(firClass, originalSymbol, contributedDeclarations)) {
+                result.add(originalSymbol.fir)
+            }
+        }
+
+        fun checkPropertySymbolAndAddToResult(originalSymbol: FirCallableSymbol<*>) {
+            if (shouldCreateFakeOverridden<FirProperty, FirPropertySymbol>(firClass, originalSymbol, contributedDeclarations)) {
+                result.add(originalSymbol.fir)
+            }
+        }
+
+        val superTypesCallableNames = classScope.getCallableNames()
+        for (name in superTypesCallableNames) {
+            classScope.processFunctionsByName(name) { functionSymbol ->
+                // TODO: MANY_* as well as some conflict diagnostics
+                //if (functionSymbol is FirIntersectionOverrideFunctionSymbol)
+                //    functionSymbol.intersections.forEach(::checkFunctionSymbolAndAddToResult)
+                //else
+                checkFunctionSymbolAndAddToResult(functionSymbol)
+            }
+
+            classScope.processPropertiesByName(name) { propertySymbol ->
+                // TODO: MANY_* as well as some conflict diagnostics
+                //if (propertySymbol is FirIntersectionOverridePropertySymbol)
+                //    propertySymbol.intersections.forEach(::checkPropertySymbolAndAddToResult)
+                //else
+                checkPropertySymbolAndAddToResult(propertySymbol)
+            }
+        }
+
+        return result
+    }
+
+    // See [FakeOverrideGenerator#createFakeOverriddenIfNeeded]
+    private inline fun <reified D : FirCallableMemberDeclaration<D>, reified S : FirCallableSymbol<D>> shouldCreateFakeOverridden(
+        firClass: FirClass<*>,
+        originalSymbol: FirCallableSymbol<*>,
+        contributedDeclarations: Collection<FirDeclaration>,
+    ): Boolean {
+        if (originalSymbol !is S || originalSymbol.fir in contributedDeclarations) return false
+        val classLookupTag = firClass.symbol.toLookupTag()
+        val originalDeclaration = originalSymbol.fir
+        if (originalSymbol.dispatchReceiverClassOrNull() == classLookupTag && !originalDeclaration.origin.fromSupertypes) return false
+        if (originalDeclaration.visibility == Visibilities.Private) return false
+        return when {
+            originalSymbol.fir.origin.fromSupertypes && originalSymbol.dispatchReceiverClassOrNull() == classLookupTag -> {
```

Comment:
The sequence of similar checks here can be simplified.

Reply:
If this will return that `boolean` result, yes, that's right. I intentionally maintain this structure to align with `FakeOverrideGenerator` since we need more accurate base symbols for substitute/intersection cases. Noticed this https://github.com/JetBrains/kotlin/commit/cf830887ec13b4b5e0d9f7b6ce46b6c6990ca1e0 too; changed the check to use that as well; and rewrote a bit to collect base symbols for substitute/intersection cases as well.